// @sapphire/framework/dist/lib/utils/logger/Logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
require("./ILogger");
class Logger {
    constructor(level) {
        this.level = level;
    }
    has(level) {
        return level >= this.level;
    }
    trace(...values) {
        this.write(10 /* LogLevel.Trace */, ...values);
    }
    debug(...values) {
        this.write(20 /* LogLevel.Debug */, ...values);
    }
    info(...values) {
        this.write(30 /* LogLevel.Info */, ...values);
    }
    warn(...values) {
        this.write(40 /* LogLevel.Warn */, ...values);
    }
    error(...values) {
        this.write(50 /* LogLevel.Error */, ...values);
    }
    fatal(...values) {
        this.write(60 /* LogLevel.Fatal */, ...values);
    }
    write(level, ...values) {
        if (!this.has(level))
            return;
        const method = Logger.levels.get(level);
        if (typeof method === 'string')
            console[method](`[${method.toUpperCase()}]`, ...values);
    }
}
exports.Logger = Logger;
Logger.levels = new Map([
    [10 /* LogLevel.Trace */, 'trace'],
    [20 /* LogLevel.Debug */, 'debug'],
    [30 /* LogLevel.Info */, 'info'],
    [40 /* LogLevel.Warn */, 'warn'],
    [50 /* LogLevel.Error */, 'error'],
    [60 /* LogLevel.Fatal */, 'error']
]);
