// @sapphire/framework/dist/lib/types/Enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InternalRegistryAPIType = exports.RegisterBehavior = exports.BucketScope = exports.PluginHook = exports.CooldownLevel = void 0;
var CooldownLevel;
(function (CooldownLevel) {
    CooldownLevel["Author"] = "author";
    CooldownLevel["Channel"] = "channel";
    CooldownLevel["Guild"] = "guild";
})(CooldownLevel = exports.CooldownLevel || (exports.CooldownLevel = {}));
var PluginHook;
(function (PluginHook) {
    PluginHook["PreGenericsInitialization"] = "preGenericsInitialization";
    PluginHook["PreInitialization"] = "preInitialization";
    PluginHook["PostInitialization"] = "postInitialization";
    PluginHook["PreLogin"] = "preLogin";
    PluginHook["PostLogin"] = "postLogin";
})(PluginHook = exports.PluginHook || (exports.PluginHook = {}));
/**
 * The scope the cooldown applies to.
 */
var BucketScope;
(function (BucketScope) {
    /**
     * Per channel cooldowns.
     */
    BucketScope[BucketScope["Channel"] = 0] = "Channel";
    /**
     * Global cooldowns.
     */
    BucketScope[BucketScope["Global"] = 1] = "Global";
    /**
     * Per guild cooldowns.
     */
    BucketScope[BucketScope["Guild"] = 2] = "Guild";
    /**
     * Per user cooldowns.
     */
    BucketScope[BucketScope["User"] = 3] = "User";
})(BucketScope = exports.BucketScope || (exports.BucketScope = {}));
var RegisterBehavior;
(function (RegisterBehavior) {
    RegisterBehavior["Overwrite"] = "OVERWRITE";
    RegisterBehavior["LogToConsole"] = "LOG_TO_CONSOLE";
})(RegisterBehavior = exports.RegisterBehavior || (exports.RegisterBehavior = {}));
var InternalRegistryAPIType;
(function (InternalRegistryAPIType) {
    InternalRegistryAPIType[InternalRegistryAPIType["ChatInput"] = 0] = "ChatInput";
    InternalRegistryAPIType[InternalRegistryAPIType["ContextMenu"] = 1] = "ContextMenu";
})(InternalRegistryAPIType = exports.InternalRegistryAPIType || (exports.InternalRegistryAPIType = {}));
