// @sapphire/framework/dist/lib/structures/Precondition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AllFlowsPrecondition = exports.Precondition = void 0;
const pieces_1 = require("@sapphire/pieces");
const result_1 = require("@sapphire/result");
const PreconditionError_1 = require("../errors/PreconditionError");
class Precondition extends pieces_1.Piece {
    constructor(context, options = {}) {
        super(context, options);
        this.position = options.position ?? null;
    }
    ok() {
        return (0, result_1.ok)();
    }
    /**
     * Constructs a {@link PreconditionError} with the precondition parameter set to `this`.
     * @param options The information.
     */
    error(options = {}) {
        return (0, result_1.err)(new PreconditionError_1.PreconditionError({ precondition: this, ...options }));
    }
    async fetchChannelFromInteraction(interaction) {
        const channel = (await interaction.client.channels.fetch(interaction.channelId, {
            cache: false,
            allowUnknownGuild: true
        }));
        return channel;
    }
}
exports.Precondition = Precondition;
class AllFlowsPrecondition extends Precondition {
}
exports.AllFlowsPrecondition = AllFlowsPrecondition;
