// @sapphire/framework/dist/lib/parsers/Result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fromAsync = exports.from = exports.isErr = exports.isOk = exports.err = exports.ok = void 0;
const utilities_1 = require("@sapphire/utilities");
function ok(x) {
    return { success: true, value: x };
}
exports.ok = ok;
function err(x) {
    return { success: false, error: x };
}
exports.err = err;
/**
 * Determines whether or not a result is an Ok.
 * @typeparam T The result's type.
 * @typeparam E The error's type.
 */
function isOk(x) {
    return x.success;
}
exports.isOk = isOk;
/**
 * Determines whether or not a result is an Err.
 * @typeparam T The result's type.
 * @typeparam E The error's type.
 */
function isErr(x) {
    return !x.success;
}
exports.isErr = isErr;
/**
 * Creates a {@link Result} out of a callback.
 * @typeparam T The result's type.
 * @typeparam E The error's type.
 */
function from(cb) {
    try {
        return ok(cb());
    }
    catch (error) {
        return err(error);
    }
}
exports.from = from;
/**
 * Creates a {@link Result} out of a promise or async callback.
 * @typeparam T The result's type.
 * @typeparam E The error's type.
 */
async function fromAsync(promiseOrCb) {
    try {
        return ok(await ((0, utilities_1.isFunction)(promiseOrCb) ? promiseOrCb() : promiseOrCb));
    }
    catch (error) {
        return err(error);
    }
}
exports.fromAsync = fromAsync;
