// @sapphire/framework/dist/lib/errors/Identifiers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Identifiers = void 0;
var Identifiers;
(function (Identifiers) {
    Identifiers["ArgsMissing"] = "argsMissing";
    Identifiers["ArgsUnavailable"] = "argsUnavailable";
    Identifiers["ArgumentBooleanError"] = "booleanError";
    Identifiers["ArgumentChannelError"] = "channelError";
    Identifiers["ArgumentDateError"] = "dateError";
    Identifiers["ArgumentDateTooEarly"] = "dateTooEarly";
    Identifiers["ArgumentDateTooFar"] = "dateTooFar";
    Identifiers["ArgumentDMChannelError"] = "dmChannelError";
    Identifiers["ArgumentFloatError"] = "floatError";
    Identifiers["ArgumentFloatTooLarge"] = "floatTooLarge";
    Identifiers["ArgumentFloatTooSmall"] = "floatTooSmall";
    Identifiers["ArgumentGuildCategoryChannelError"] = "categoryChannelError";
    Identifiers["ArgumentGuildChannelError"] = "guildChannelError";
    Identifiers["ArgumentGuildChannelMissingGuildError"] = "guildChannelMissingGuildError";
    Identifiers["ArgumentGuildNewsChannelError"] = "guildNewsChannelError";
    Identifiers["ArgumentGuildNewsThreadChannelError"] = "guildNewsChannelError";
    Identifiers["ArgumentGuildPrivateThreadChannelError"] = "guildPrivateThreadChannelError";
    Identifiers["ArgumentGuildPublicThreadChannelError"] = "guildPublicThreadChannelError";
    Identifiers["ArgumentGuildStageVoiceChannelError"] = "guildStageVoiceChannelError";
    Identifiers["ArgumentGuildTextChannelError"] = "guildTextChannelError";
    Identifiers["ArgumentGuildThreadChannelError"] = "guildThreadChannelError";
    Identifiers["ArgumentGuildVoiceChannelError"] = "guildVoiceChannelError";
    Identifiers["ArgumentHyperlinkError"] = "hyperlinkError";
    Identifiers["ArgumentIntegerError"] = "integerError";
    Identifiers["ArgumentIntegerTooLarge"] = "integerTooLarge";
    Identifiers["ArgumentIntegerTooSmall"] = "integerTooSmall";
    Identifiers["ArgumentMemberError"] = "memberError";
    Identifiers["ArgumentMemberMissingGuild"] = "memberMissingGuild";
    Identifiers["ArgumentMessageError"] = "messageError";
    Identifiers["ArgumentNumberError"] = "numberError";
    Identifiers["ArgumentNumberTooLarge"] = "numberTooLarge";
    Identifiers["ArgumentNumberTooSmall"] = "numberTooSmall";
    Identifiers["ArgumentRoleError"] = "roleError";
    Identifiers["ArgumentRoleMissingGuild"] = "roleMissingGuild";
    Identifiers["ArgumentStringTooLong"] = "stringTooLong";
    Identifiers["ArgumentStringTooShort"] = "stringTooShort";
    Identifiers["ArgumentUserError"] = "userError";
    Identifiers["ArgumentEnumEmptyError"] = "enumEmptyError";
    Identifiers["ArgumentEnumError"] = "enumError";
    Identifiers["CommandDisabled"] = "commandDisabled";
    Identifiers["PreconditionCooldown"] = "preconditionCooldown";
    Identifiers["PreconditionDMOnly"] = "preconditionDmOnly";
    Identifiers["PreconditionGuildNewsOnly"] = "preconditionGuildNewsOnly";
    Identifiers["PreconditionGuildNewsThreadOnly"] = "preconditionGuildNewsThreadOnly";
    Identifiers["PreconditionGuildOnly"] = "preconditionGuildOnly";
    Identifiers["PreconditionGuildPrivateThreadOnly"] = "preconditionGuildPrivateThreadOnly";
    Identifiers["PreconditionGuildPublicThreadOnly"] = "preconditionGuildPublicThreadOnly";
    Identifiers["PreconditionGuildTextOnly"] = "preconditionGuildTextOnly";
    Identifiers["PreconditionNSFW"] = "preconditionNsfw";
    Identifiers["PreconditionClientPermissions"] = "preconditionClientPermissions";
    Identifiers["PreconditionClientPermissionsNoClient"] = "preconditionClientPermissionsNoClient";
    Identifiers["PreconditionClientPermissionsNoPermissions"] = "preconditionClientPermissionsNoPermissions";
    Identifiers["PreconditionUserPermissions"] = "preconditionUserPermissions";
    Identifiers["PreconditionUserPermissionsNoPermissions"] = "preconditionUserPermissionsNoPermissions";
    Identifiers["PreconditionThreadOnly"] = "preconditionThreadOnly";
})(Identifiers = exports.Identifiers || (exports.Identifiers = {}));
