// lexure/dist/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.args = require("./args");
exports.lexer = require("./lexer");
exports.loops = require("./loops");
exports.loopAction = require("./loopAction");
exports.option = require("./option");
exports.parser = require("./parser");
exports.parserOutput = require("./parserOutput");
exports.result = require("./result");
exports.tokens = require("./tokens");
exports.unordered = require("./unordered");
exports.util = require("./util");
__exportStar(require("./args"), exports);
__exportStar(require("./lexer"), exports);
__exportStar(require("./loops"), exports);
__exportStar(require("./loopAction"), exports);
__exportStar(require("./option"), exports);
__exportStar(require("./parser"), exports);
__exportStar(require("./parserOutput"), exports);
__exportStar(require("./result"), exports);
__exportStar(require("./tokens"), exports);
__exportStar(require("./unordered"), exports);
__exportStar(require("./util"), exports);
