// lexure/dist/lib/loopAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fail_ = exports.fail = exports.finish = exports.step_ = exports.step = exports.LoopTag = void 0;
/**
 * Tag for the loop action variants.
 */
var LoopTag;
(function (LoopTag) {
    LoopTag["STEP"] = "step";
    LoopTag["FINISH"] = "finish";
    LoopTag["FAIL"] = "fail";
})(LoopTag = exports.LoopTag || (exports.LoopTag = {}));
/**
 * Creates a Step.
 * @typeparam A - Type of step results.
 * @param x - Value to use.
 * @returns A LoopAction.
 */
function step(x) {
    return { action: LoopTag.STEP, item: x };
}
exports.step = step;
/**
 * Creates a Step with null value.
 * @returns A LoopAction.
 */
function step_() {
    return { action: LoopTag.STEP, item: null };
}
exports.step_ = step_;
/**
 * Creates a Finish.
 * @typeparam B - Type of finish results.
 * @param x - Value to use.
 * @returns A LoopAction.
 */
function finish(x) {
    return { action: LoopTag.FINISH, value: x };
}
exports.finish = finish;
/**
 * Creates a Fail.
 * @typeparam E - Type of errors.
 * @param x - Value to use.
 * @returns A LoopAction.
 */
function fail(x) {
    return { action: LoopTag.FAIL, error: x };
}
exports.fail = fail;
/**
 * Creates a Fail with null value.
 * @returns A LoopAction.
 */
function fail_() {
    return { action: LoopTag.FAIL, error: null };
}
exports.fail_ = fail_;
