// @sapphire/framework/dist/preconditions/GuildNewsOnly.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CorePrecondition = void 0;
require("../lib/errors/Identifiers");
const Precondition_1 = require("../lib/structures/Precondition");
class CorePrecondition extends Precondition_1.AllFlowsPrecondition {
    constructor() {
        super(...arguments);
        this.allowedTypes = ['GUILD_NEWS', 'GUILD_NEWS_THREAD'];
    }
    messageRun(message) {
        return this.allowedTypes.includes(message.channel.type)
            ? this.ok()
            : this.error({
                identifier: "preconditionGuildNewsOnly" /* Identifiers.PreconditionGuildNewsOnly */,
                message: 'You can only run this message command in server announcement channels.'
            });
    }
    async chatInputRun(interaction) {
        const channel = await this.fetchChannelFromInteraction(interaction);
        return this.allowedTypes.includes(channel.type)
            ? this.ok()
            : this.error({
                identifier: "preconditionGuildNewsOnly" /* Identifiers.PreconditionGuildNewsOnly */,
                message: 'You can only run this chat input command in server announcement channels.'
            });
    }
    async contextMenuRun(interaction) {
        const channel = await this.fetchChannelFromInteraction(interaction);
        return this.allowedTypes.includes(channel.type)
            ? this.ok()
            : this.error({
                identifier: "preconditionGuildNewsOnly" /* Identifiers.PreconditionGuildNewsOnly */,
                message: 'You can only run this context menu command in server announcement channels.'
            });
    }
}
exports.CorePrecondition = CorePrecondition;
