// @sapphire/framework/dist/preconditions/GuildPublicThreadOnly.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CorePrecondition = void 0;
require("../lib/errors/Identifiers");
const Precondition_1 = require("../lib/structures/Precondition");
class CorePrecondition extends Precondition_1.AllFlowsPrecondition {
    messageRun(message) {
        return message.thread?.type === 'GUILD_PUBLIC_THREAD'
            ? this.ok()
            : this.error({
                identifier: "preconditionGuildPublicThreadOnly" /* Identifiers.PreconditionGuildPublicThreadOnly */,
                message: 'You can only run this message command in public server thread channels.'
            });
    }
    async chatInputRun(interaction) {
        const channel = await this.fetchChannelFromInteraction(interaction);
        return channel.type === 'GUILD_PUBLIC_THREAD'
            ? this.ok()
            : this.error({
                identifier: "preconditionGuildPublicThreadOnly" /* Identifiers.PreconditionGuildPublicThreadOnly */,
                message: 'You can only run this chat input command in public server thread channels.'
            });
    }
    async contextMenuRun(interaction) {
        const channel = await this.fetchChannelFromInteraction(interaction);
        return channel.type === 'GUILD_PUBLIC_THREAD'
            ? this.ok()
            : this.error({
                identifier: "preconditionGuildPublicThreadOnly" /* Identifiers.PreconditionGuildPublicThreadOnly */,
                message: 'You can only run this context menu command in public server thread channels.'
            });
    }
}
exports.CorePrecondition = CorePrecondition;
