// @sapphire/framework/dist/preconditions/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserPermissions = exports.NSFW = exports.GuildThreadOnly = exports.GuildTextOnly = exports.GuildPublicThreadOnly = exports.GuildPrivateThreadOnly = exports.GuildOnly = exports.GuildNewsThreadOnly = exports.GuildNewsOnly = exports.Enabled = exports.DMOnly = exports.Cooldown = exports.ClientPermissions = void 0;
var ClientPermissions_1 = require("./ClientPermissions");
Object.defineProperty(exports, "ClientPermissions", { enumerable: true, get: function () { return ClientPermissions_1.CorePrecondition; } });
var Cooldown_1 = require("./Cooldown");
Object.defineProperty(exports, "Cooldown", { enumerable: true, get: function () { return Cooldown_1.CorePrecondition; } });
var DMOnly_1 = require("./DMOnly");
Object.defineProperty(exports, "DMOnly", { enumerable: true, get: function () { return DMOnly_1.CorePrecondition; } });
var Enabled_1 = require("./Enabled");
Object.defineProperty(exports, "Enabled", { enumerable: true, get: function () { return Enabled_1.CorePrecondition; } });
var GuildNewsOnly_1 = require("./GuildNewsOnly");
Object.defineProperty(exports, "GuildNewsOnly", { enumerable: true, get: function () { return GuildNewsOnly_1.CorePrecondition; } });
var GuildNewsThreadOnly_1 = require("./GuildNewsThreadOnly");
Object.defineProperty(exports, "GuildNewsThreadOnly", { enumerable: true, get: function () { return GuildNewsThreadOnly_1.CorePrecondition; } });
var GuildOnly_1 = require("./GuildOnly");
Object.defineProperty(exports, "GuildOnly", { enumerable: true, get: function () { return GuildOnly_1.CorePrecondition; } });
var GuildPrivateThreadOnly_1 = require("./GuildPrivateThreadOnly");
Object.defineProperty(exports, "GuildPrivateThreadOnly", { enumerable: true, get: function () { return GuildPrivateThreadOnly_1.CorePrecondition; } });
var GuildPublicThreadOnly_1 = require("./GuildPublicThreadOnly");
Object.defineProperty(exports, "GuildPublicThreadOnly", { enumerable: true, get: function () { return GuildPublicThreadOnly_1.CorePrecondition; } });
var GuildTextOnly_1 = require("./GuildTextOnly");
Object.defineProperty(exports, "GuildTextOnly", { enumerable: true, get: function () { return GuildTextOnly_1.CorePrecondition; } });
var GuildThreadOnly_1 = require("./GuildThreadOnly");
Object.defineProperty(exports, "GuildThreadOnly", { enumerable: true, get: function () { return GuildThreadOnly_1.CorePrecondition; } });
var NSFW_1 = require("./NSFW");
Object.defineProperty(exports, "NSFW", { enumerable: true, get: function () { return NSFW_1.CorePrecondition; } });
var UserPermissions_1 = require("./UserPermissions");
Object.defineProperty(exports, "UserPermissions", { enumerable: true, get: function () { return UserPermissions_1.CorePrecondition; } });
