// @sapphire/result/dist/index.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/lib/Maybe.ts
function maybe(value) {
  return isMaybe(value) ? value : value === null ? none() : some(value);
}
function some(x) {
  return { exists: true, value: x };
}
function none() {
  return { exists: false };
}
function isSome(x) {
  return x.exists;
}
function isNone(x) {
  return !x.exists;
}
function isMaybe(x) {
  return typeof x === "object" && x !== null && typeof Reflect.get(x, "exists") === "boolean";
}

// src/lib/Result.ts
function ok(x) {
  return { success: true, value: x };
}
function err(x) {
  return { success: false, error: x };
}
function isOk(x) {
  return x.success;
}
function isErr(x) {
  return !x.success;
}

// src/lib/Parsers.ts
function from(cb) {
  try {
    return ok(cb());
  } catch (error) {
    return err(error);
  }
}
async function fromAsync(promiseOrCb) {
  try {
    return ok(await (isFunction(promiseOrCb) ? promiseOrCb() : promiseOrCb));
  } catch (error) {
    return err(error);
  }
}
function isFunction(input) {
  return typeof input === "function";
}
export {
  err,
  from,
  fromAsync,
  isErr,
  isMaybe,
  isNone,
  isOk,
  isSome,
  maybe,
  none,
  ok,
  some
};
