// @sapphire/result/dist/index.global.js
var SapphireResult = (() => {
  var __defProp = Object.defineProperty;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
  var __export = (target, all) => {
    for (var name in all)
      __defProp(target, name, { get: all[name], enumerable: true });
  };
  var __copyProps = (to, from2, except, desc) => {
    if (from2 && typeof from2 === "object" || typeof from2 === "function") {
      for (let key of __getOwnPropNames(from2))
        if (!__hasOwnProp.call(to, key) && key !== except)
          __defProp(to, key, { get: () => from2[key], enumerable: !(desc = __getOwnPropDesc(from2, key)) || desc.enumerable });
    }
    return to;
  };
  var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

  // src/index.ts
  var src_exports = {};
  __export(src_exports, {
    err: () => err,
    from: () => from,
    fromAsync: () => fromAsync,
    isErr: () => isErr,
    isMaybe: () => isMaybe,
    isNone: () => isNone,
    isOk: () => isOk,
    isSome: () => isSome,
    maybe: () => maybe,
    none: () => none,
    ok: () => ok,
    some: () => some
  });

  // src/lib/Maybe.ts
  function maybe(value) {
    return isMaybe(value) ? value : value === null ? none() : some(value);
  }
  function some(x) {
    return { exists: true, value: x };
  }
  function none() {
    return { exists: false };
  }
  function isSome(x) {
    return x.exists;
  }
  function isNone(x) {
    return !x.exists;
  }
  function isMaybe(x) {
    return typeof x === "object" && x !== null && typeof Reflect.get(x, "exists") === "boolean";
  }

  // src/lib/Result.ts
  function ok(x) {
    return { success: true, value: x };
  }
  function err(x) {
    return { success: false, error: x };
  }
  function isOk(x) {
    return x.success;
  }
  function isErr(x) {
    return !x.success;
  }

  // src/lib/Parsers.ts
  function from(cb) {
    try {
      return ok(cb());
    } catch (error) {
      return err(error);
    }
  }
  async function fromAsync(promiseOrCb) {
    try {
      return ok(await (isFunction(promiseOrCb) ? promiseOrCb() : promiseOrCb));
    } catch (error) {
      return err(error);
    }
  }
  function isFunction(input) {
    return typeof input === "function";
  }
  return __toCommonJS(src_exports);
})();
