// @sapphire/pieces/dist/lib/structures/StoreRegistry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StoreRegistry = void 0;
const tslib_1 = require("tslib");
const collection_1 = tslib_1.__importDefault(require("@discordjs/collection"));
const path_1 = require("path");
const RootScan_1 = require("../internal/RootScan");
/**
 * A strict-typed store registry. This is available in {@link container}.
 * @since 2.1.0
 * @example
 * ```typescript
 * // Adding new stores
 *
 * // Register the store:
 * container.stores.register(new RouteStore());
 *
 * // Augment Sapphire to add the new store, in case of a JavaScript
 * // project, this can be moved to an `Augments.d.ts` (or any other name)
 * // file somewhere:
 * declare module '@sapphire/pieces' {
 *   export interface StoreRegistryEntries {
 *     routes: RouteStore;
 *   }
 * }
 * ```
 */
class StoreRegistry extends collection_1.default {
    /**
     * Loads all the registered stores.
     * @since 2.1.0
     */
    async load() {
        const promises = [];
        for (const store of this.values()) {
            promises.push(store.loadAll());
        }
        await Promise.all(promises);
    }
    /**
     * Registers all user directories from the process working directory, the default value is obtained by assuming
     * CommonJS (high accuracy) but with fallback for ECMAScript Modules (reads package.json's `main` entry, fallbacks
     * to `process.cwd()`).
     *
     * By default, if you have this folder structure:
     * ```
     * /home/me/my-bot
     * ├─ src
     * │  ├─ commands
     * │  ├─ events
     * │  └─ main.js
     * └─ package.json
     * ```
     *
     * And you run `node src/main.js`, the directories `/home/me/my-bot/src/commands` and `/home/me/my-bot/src/events` will
     * be registered for the commands and events stores respectively, since both directories are located in the same
     * directory as your main file.
     *
     * **Note**: this also registers directories for all other stores, even if they don't have a folder, this allows you
     * to create new pieces and hot-load them later anytime.
     * @since 2.1.0
     * @param rootDirectory The root directory to register pieces at.
     */
    registerPath(rootDirectory = (0, RootScan_1.getRootData)().root) {
        for (const store of this.values()) {
            store.registerPath((0, path_1.join)(rootDirectory, store.name));
        }
    }
    /**
     * Registers a store.
     * @since 2.1.0
     * @param store The store to register.
     */
    register(store) {
        this.set(store.name, store);
        return this;
    }
    /**
     * Deregisters a store.
     * @since 2.1.0
     * @param store The store to deregister.
     */
    deregister(store) {
        this.delete(store.name);
        return this;
    }
}
exports.StoreRegistry = StoreRegistry;
