// @sapphire/framework/dist/lib/utils/application-commands/normalizeInputs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertApplicationCommandToApiData = exports.normalizeContextMenuCommand = exports.normalizeChatInputCommand = void 0;
const builders_1 = require("@discordjs/builders");
const utilities_1 = require("@sapphire/utilities");
const v9_1 = require("discord-api-types/v9");
const discord_js_1 = require("discord.js");
function isBuilder(command) {
    return command instanceof builders_1.SlashCommandBuilder;
}
function normalizeChatInputCommand(command) {
    if ((0, utilities_1.isFunction)(command)) {
        const builder = new builders_1.SlashCommandBuilder();
        command(builder);
        return builder.toJSON();
    }
    if (isBuilder(command)) {
        return command.toJSON();
    }
    const finalObject = {
        name: command.name,
        name_localizations: command.nameLocalizations,
        description: command.description,
        description_localizations: command.descriptionLocalizations,
        default_permission: command.defaultPermission,
        type: v9_1.ApplicationCommandType.ChatInput
        // TODO: once command perms v2 drops, add the fields here
    };
    if (command.options?.length) {
        finalObject.options = command.options.map((option) => discord_js_1.ApplicationCommand['transformOption'](option));
    }
    return finalObject;
}
exports.normalizeChatInputCommand = normalizeChatInputCommand;
function normalizeContextMenuCommand(command) {
    if ((0, utilities_1.isFunction)(command)) {
        const builder = new builders_1.ContextMenuCommandBuilder();
        command(builder);
        return builder.toJSON();
    }
    if (command instanceof builders_1.ContextMenuCommandBuilder) {
        return command.toJSON();
    }
    let type;
    switch (command.type) {
        case discord_js_1.Constants.ApplicationCommandTypes.MESSAGE:
        case 'MESSAGE':
            type = v9_1.ApplicationCommandType.Message;
            break;
        case discord_js_1.Constants.ApplicationCommandTypes.USER:
        case 'USER':
            type = v9_1.ApplicationCommandType.User;
            break;
        default:
            // @ts-expect-error command gets turned to never, which is half true.
            throw new Error(`Unhandled command type: ${command.type}`);
    }
    const finalObject = {
        name: command.name,
        name_localizations: command.nameLocalizations,
        type,
        default_permission: command.defaultPermission
        // TODO: once command perms v2 drops, add the fields here
    };
    return finalObject;
}
exports.normalizeContextMenuCommand = normalizeContextMenuCommand;
function convertApplicationCommandToApiData(command) {
    const returnData = {
        name: command.name,
        name_localizations: command.nameLocalizations,
        default_permission: command.defaultPermission
        // TODO: once command perms v2 drops, add the fields here
    };
    if (command.type === 'CHAT_INPUT') {
        returnData.type = v9_1.ApplicationCommandType.ChatInput;
        returnData.description = command.description;
        // TODO (favna): Remove this line after website rewrite is done
        // @ts-ignore this is currently ignored for the website
        returnData.description_localizations = command.descriptionLocalizations;
    }
    else if (command.type === 'MESSAGE') {
        returnData.type = v9_1.ApplicationCommandType.Message;
    }
    else if (command.type === 'USER') {
        returnData.type = v9_1.ApplicationCommandType.User;
    }
    if (command.options.length) {
        returnData.options = command.options.map((option) => discord_js_1.ApplicationCommand['transformOption'](option));
    }
    return returnData;
}
exports.convertApplicationCommandToApiData = convertApplicationCommandToApiData;
